aten/src/ATen/native/cuda/UpSampleBilinear2d.cu

```diff
@@ -234,35 +384,70 @@ static void upsample_bilinear2d_backward_out_cuda_template(
       at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, 1024);
   cudaStream_t stream = at::cuda::getCurrentCUDAStream();
 
-  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
-      grad_output.scalar_type(), "upsample_bilinear2d_backward_out_frame", [&] {
-        using accscalar_t = at::acc_type<scalar_t, true>;
-
-        auto idata = grad_input.data_ptr<scalar_t>();
-        auto odata = grad_output.data_ptr<scalar_t>();
-
-        const accscalar_t rheight = area_pixel_compute_scale<accscalar_t>(
-            input_height, output_height, align_corners, scales_h);
-        const accscalar_t rwidth = area_pixel_compute_scale<accscalar_t>(
-            input_width, output_width, align_corners, scales_w);
-
-        upsample_bilinear2d_backward_out_frame<scalar_t, accscalar_t>
-            <<<cuda::ATenCeilDiv(num_kernels, static_cast<size_t>(num_threads)),
-               num_threads,
-               0,
-               stream>>>(
-                nbatch * channels,
-                input_height,
-                input_width,
-                output_height,
-                output_width,
-                rheight,
-                rwidth,
-                align_corners,
-                idata,
-                odata);
-        C10_CUDA_KERNEL_LAUNCH_CHECK();
-      });
+  if (grad_output.sizes() == grad_input.sizes()) {
+    grad_input.copy_(grad_output);
+    return;
+  }
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(grad_output.scalar_type(), "upsample_bilinear2d_backward_out_frame", [&] {
+    if (memory_format == at::MemoryFormat::ChannelsLast) {
```

Comment:
is channels-last backward always faster?

Reply:
@ngimel For UpSampleBilinear2d.cu, 1024 threads does not actually spill. I tested by making a dummy main.cu that does nothing but `#include <ATen/native/cuda/UpSampleBilinear2d.cu>` and ran the following nvcc command to compile with the --resource-usage flag:

`nvcc main.cu  --resource-usage --extended-lambda --expt-relaxed-constexpr \
-gencode arch=compute_70,code=sm_70 \
-DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ \
-Ipytorch -Ipytorch/aten/src/ -Ipytorch/build -Ipytorch/build/aten/src -Ipytorch/build/caffe2/aten/src \
-Lpytorch/build/lib/ -Lpytorch/build/lib.linux-x86_64-3.6/torch/lib \
-lc10 -lc10_cuda -lc10d_cuda_test -ltorch -ltorch_python -ltorch_global_deps -ltorch_cpu -ltorch_cuda -ltorchbind_test`

Reply:
Thanks for checking! Interesting, the kernel looks like it would use a lot of registers.

Reply:
Thanks for the check @Fuzzkatt and comments @ngimel . Currently on master branch, the backward always gives contiguous `grad_input`. I've added the functionality to first write backward grad_input to a temporary tensor and then copy to the actual output. With this, the backward would also give channels_last `grad_input` if needed. I re-evaluate the channels_last backward heuristics against the new approach.

Reply:
For clarity, with your changes grad_input is now allocated as channels-last if grad_output is channels-last, right? In some cases that would lead to a copy, but for the simple cases layout would be propagated without additional copies, IIUC.

Reply:
Right, `grad_input` is allocated as channels-last if `grad_output` is channels-last.

Because of the requested heuristics, some cases would use the contiguous path for computation, and the result is copied to channels-last `grad_input` at the end.

All channels-last `grad_output` will lead to channels-last `grad_input`.